domain/credential/state/state.go

```diff
@@ -507,3 +509,74 @@ func (st *State) WatchCredential(
 	result, err := getWatcher("cloud_credential", uuid, changestream.All)
 	return result, errors.Annotatef(err, "watching credential")
 }
+
+// GetCloud returns the cloud with the specified name.
+func (st *State) GetCloud(ctx context.Context, name string) (cloud.Cloud, error) {
+	db, err := st.DB()
+	if err != nil {
+		return cloud.Cloud{}, errors.Trace(err)
+	}
+
+	var result []cloud.Cloud
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		var err error
+		result, err = dbcloud.LoadClouds(ctx, tx, name)
+		return errors.Trace(err)
+	})
+	if len(result) == 0 {
+		return cloud.Cloud{}, fmt.Errorf("cloud %q %w", name, errors.NotFound)
+	}
+	return result[0], errors.Trace(err)
+}
+
+// ModelsUsingCloudCredential returns a map of uuid->name for models which use the credential.
+func (st *State) ModelsUsingCloudCredential(ctx context.Context, id credential.ID) (map[model.UUID]string, error) {
+	if err := id.Validate(); err != nil {
+		return nil, errors.Annotate(err, "invalid credential querying models")
+	}
+
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	query := `
+SELECT mm.model_uuid AS &M.model_uuid, mm.name AS &M.name
+FROM   model_metadata mm
+JOIN cloud_credential cc ON cc.uuid = mm.cloud_credential_uuid
+JOIN cloud ON cloud.uuid = cc.cloud_uuid
+`
+
+	types := []any{
+		sqlair.M{},
+	}
+	condition, args := database.SqlairClauseAnd(map[string]any{
+		"cc.name":       id.Name,
+		"cloud.name":    id.Cloud,
+		"cc.owner_uuid": id.Owner,
+	})
+	query = query + "WHERE " + condition
+
+	stmt, err := sqlair.Prepare(query, types...)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	var info []sqlair.M
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, stmt, args).GetAll(&info)
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(info) == 0 {
+		return nil, nil
```

Comment:
This should return a safe pointer IMO or document that fact that it's valid for err to be nil and the map to be nil.

Reply:
Actually why not just get rid of this if statement all together ?

Reply:
We use this pattern elsewhere for map and slice returns - the caller just iterates over the result anyway. Can change here but that still leaves the other places.

Reply:
Sorry I didn't realise you can iterate over nil. That is fine then but still makes it safer and less code to just delete. Same result just slightly safer.